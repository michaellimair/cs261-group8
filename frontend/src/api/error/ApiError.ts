class ApiError<T> extends Error {
  data?: T;

  constructor(message: string, data?: T) {
    super(message ?? 'An error has occurred in the API.');
    this.data = data;
  }

  static isApiError(obj: any): obj is ApiError<any> {
    return obj instanceof ApiError;
  }
}

export default ApiError;
