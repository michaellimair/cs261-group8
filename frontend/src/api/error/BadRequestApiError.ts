import ApiError from './ApiError';

interface ValidationErrorMeta<T> {
  property: keyof T,
  value: T[keyof T];
}

class BadRequestApiError<T = Record<string, string>> extends ApiError<ValidationErrorMeta<T>[]> {
  static defaultMessage = 'Validation error in the data.';

  constructor(meta: ValidationErrorMeta<T>[]) {
    super(BadRequestApiError.defaultMessage, meta);
  }

  static isBadRequestApiError<Res>(obj: any): obj is BadRequestApiError<Res> {
    return obj instanceof BadRequestApiError;
  }
}

export default BadRequestApiError;
