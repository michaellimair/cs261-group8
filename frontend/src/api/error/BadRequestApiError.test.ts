import ApiError from './ApiError';
import BadRequestApiError from './BadRequestApiError';

describe('BadRequestApiError', () => {
  it('is instantiated successfully', () => {
    interface ISampleError {
      status_code: number;
      message: string;
    }

    const errorData = {
      status_code: 400,
      message: 'An error has occurred',
      errors: [],
    };

    const error = new BadRequestApiError<ISampleError>('An error has occurred', errorData);

    expect(error.data).toMatchObject(errorData);
    expect(error).toMatchSnapshot();
  });

  it('is instantiated successfully with no message', () => {
    const err = new BadRequestApiError();

    expect(err.data).toBeUndefined();
    expect(err.message).toBe(BadRequestApiError.defaultMessage);
  });

  it('is a subclass of ApiError', () => {
    const err = new BadRequestApiError<unknown>('', []);

    expect(ApiError.isApiError(err)).toBe(true);
  });

  describe('isBadRequestApiError', () => {
    it('identifies BadRequestApiError classes correctly', () => {
      const err = new BadRequestApiError<undefined>('');

      expect(BadRequestApiError.isBadRequestApiError(err)).toBe(true);
    });
  });
});
