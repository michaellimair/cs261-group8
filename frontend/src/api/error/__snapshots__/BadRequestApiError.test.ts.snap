// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadRequestApiError is instantiated successfully 1`] = `[Error: Validation error in the data.]`;
